sdks/python/apache_beam/transforms/util.py

```diff
@@ -1644,3 +1645,39 @@ def _process(element):
       yield r
 
     return pcoll | FlatMap(_process)
+
+
+@typehints.with_input_types(T)
+@typehints.with_output_types(T)
+class WaitOn(PTransform):
+  """Delays processing of a {@link PCollection} until another set of
+  PCollections has finished being processed. For example::
+
+     X | WaitOn(Y, Z) | SomeTransform()
+
+  would ensure that PCollections Y and Z (and hence their producing transforms)
+  are complete before SomeTransform gets executed on the elements of X.
+  This can be especially useful the waited-on PCollections are the outputs
+  of transforms that interact with external systems (such as writing to a
+  database or other sink).
+
+  For streaming, this delay is done on a per-window basis, i.e.
+  the corresponding window of each waited-on PCollection is computed before
+  elements are passed through the main collection.
+
+  This barrier often induces a fusion break.
+  """
+  def __init__(self, *to_be_waited_on):
+    self._to_be_waited_on = to_be_waited_on
+
+  def expand(self, pcoll):
+    # All we care about is the watermark, not the data itself.
+    # The GroupByKey avoids writing empty files for each shard, and also
+    # ensures the respective window finishes before advancing the timestamp.
+    sides = [
+        pvalue.AsIter(
+            pcoll
+            | f"WaitFor{ix}" >> (beam.FlatMap(lambda x: ()) | GroupByKey()))
```

Comment:
Interesting, so given that your test passes, we don't need to make sure that transform apply tags `WaitFor{ix}` are unique when WaitOn is used multiple times? I thought those should be unique within the entire  pipeline.

(nit) not sure if you intentionally use  WaitFor instead of WaitOn.

Reply:
Good call. Done.